### DISCUSSION TITLE
How to see what columns is derferred on query?

### DISCUSSION THREAD

**Discussion by gomeslucasm** [sqlalchemy/sqlalchemy#8361] [Usage Questions]
I'm using a field in a filter but defering it in the options, I would like it not to be loaded. My problem is that this data is used for the filter, but it shouldn't be sent to most users.

```python
data = db.session.query(News).filter(News.to.any(current_user.id)).options(defer(News.to)).all()
``` 
When i run this query and access `instance_state(data).unloaded` it not shows `['to']`. My bet is that as I use the field in the filter, sqlalchemy loads it and removes from `unloaded`, with good reason.

I know I could do this by specifying the fields in the query, 

```python
data = db.session.query(News.id, News.content, ...).filter(News.to.any(current_user.id)).options(defer(News.to)).all()
```
but I would like to know if there is any way to check for defered fields.

**Comment by zzzeek** [comment]
there's no public API to check that information up front given a select() or Query() object.   the information is there of course but the compilation process for the Query would need to proceed so that the loader options set themselves up, then there's an internal structure that represents what "paths" will receive what loader strategies, as far as the options that were given in the query structure.   this still doesn't take into account the class-configured loader strategies either, still another pass would be needed to resolve all of that.

so from a practicality perspective, the two immediate ways to just get an idea of this are 1. stringify the query and look at the columns or 2. load a few objects and see what's loaded in them, and that part we do have an API which you can see [here](https://docs.sqlalchemy.org/en/14/orm/internals.html?highlight=instancestate#sqlalchemy.orm.AttributeState.loaded_value), such as:

```py
from sqlalchemy import inspect
from sqlalchemy.orm.attributes import NO_VALUE

def to_is_loaded(some_object):
   insp = inspect(some_object)
   return insp.attrs.to.loaded_value is not NO_VALUE
```
